src/pages/blog/2024-08-14-semantic-nullability-for-application-developers.mdx

```diff
@@ -0,0 +1,93 @@
+
+This results in `null` having two different meanings in GraphQL.
+1. No value was provided. The `User` never provided an `age`.
+2. There was an error resolving the field. The `User` provided an age in the form of a `Float` and it couldn't be cast to an `Int`, or our new `BirthdayBoy` provider [microservice](https://www.youtube.com/watch?v=y8OnoxKotPQ) timed out and never returned the `User`'s birthday.
+
+We can differentiate between the two by calling the first `null`, and the second `(Error, null)`. The first type is signaled by a `null` return value and no associated error in the errors array. The second is signaled by a `null` return value and an associated error in the errors array.
+
+Returning to the GraphQL type system, we can see it has two options to indicate the nullability of a field.
+1. `String` which we now know means the field can be a `String`, `null`, or `(Error, null)`
+2. `String!` which we now know means the field can be `String`
+
+When an error occurs resolving a `String` field, it's not much of a problem for clients. They can decide if they can deal with that field missing or not. However when an error occurs resolving a `String!` field, GraphQL responds by destroying part of the result data before it's sent to the client. Given the danger, many developers choose to never use any non-nullable fields.
+
+Every time developers are surveyed about their issues with GraphQL, they talk about nullability. The Nullability Working Group has been hard at work, and we believe we finally have a solution.
+
+The root of the problem is that developers want a way to express that all `User`s are expected to have an `age`, but if there is an error and GraphQL can't resolve `age`, then they'd like to deal with it client-side. In order to do that, they need the type system to allow for a third type of nullability.
+
+3. A field which can be `String` or `(Error, null)`
+
+This is what we're calling "Semantic non-null". The syntax we've chosen is the following
+
+| Syntax  | Meaning           |
+| ------- | ----------------- |
+| String? | Nullable          |
+| String  | Semantic non-null |
+| String! | Strict non-null   |
+
+Types are now semantic non-null by default. Question marks are used to indicate a nullable field similar to many other modern languages. `String!` retains its meaning. This is of course, a breaking change, and GraphQL prides itself in offering a path to non-breaking evolution for existing services. So alongside the new type, we're introducing some mechanics to assist developers in making incremental updates to their applications.
```

Comment:
Rather than saying this _is_ a breaking change, let's say it _would be_ a breaking change and then we can clarify why it isn't.

```suggestion
With this syntax, types become semantic non-null by default. Question marks (`?`) are used to indicate a nullable field similar to many other modern languages. `String!` retains its meaning.

This would be a breaking change; but GraphQL prides itself in offering a path to non-breaking evolution for existing services, so alongside the new type we're introducing some mechanics to assist developers in making incremental updates to their applications.
```